# Actividad - Proyecto práctico


> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.
*   Alumno 1: Ane Moreno Oya
*   Alumno 2: Isabel María Suárez Sánchez
*   Alumno 3: Naroa Alonso Fernández
*   Alumno 4: Antonio Manuel Míguez Vega






---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [ ]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/drive/'
drive_root = mount + "/My Drive/08_MIAR/actividades/proyecto practico/"

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False

---
### 1.3. Montar carpeta de datos local (solo Colab)

In [ ]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

We're running Colab
Colab: mounting Google drive on  /content/drive/
Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).

Colab: making sure  /content/drive//My Drive/08_MIAR/actividades/proyecto practico/  exists.

Colab: Changing directory to  /content/drive//My Drive/08_MIAR/actividades/proyecto practico/
/content/drive/.shortcut-targets-by-id/1bjlA0Texh0AAABSiB7sYbV94utLfnekZ/proyecto practico
Archivos en el directorio: 
['.ipynb_checkpoints', 'dqn_vanilla_baseline', 'dqn_vanilla_baseline_finetune', 'dqn_double_mem200k', 'dqn_vanilla_lr5', 'dqn_vanilla_smallnet', 'dqn_vanilla_mem200k', 'dqn_vanilla_eps250k', 'dqn_double_eps250k', 'dqn_double_lr5', 'dqn_double_baseline', 'dqn_double_bs_64', 'dqn_vanilla_baseline_SpaceInvaders-v0_weights.h5f.data-00000-of-00001', 'dqn_vanilla_baseline_SpaceInvaders-v0_weights.h5f.index', 'dqn_double_lr05_SpaceInvaders-v0_weights.h5f.index', 'dqn_double_lr05_SpaceInvaders-v

---
### 1.4. Instalar librerías necesarias

In [ ]:
if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.12.0
  %pip install pyvirtualdisplay
else:
  %pip install pyvirtualdisplay
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.5.3
  %pip install torch==2.0.1
  %pip install agents==1.4.0

  Cloning https://github.com/Kojoley/atari-py.git to /tmp/pip-req-build-rdc4fo_y
  Running command git clone --filter=blob:none --quiet https://github.com/Kojoley/atari-py.git /tmp/pip-req-build-rdc4fo_y
  Resolved https://github.com/Kojoley/atari-py.git to commit 86a1e05c0a95e9e6233c3a413521fdb34ca8a089
  Preparing metadata (setup.py) ... done


## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, el requisito mínimo será alcanzado cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

**Rúbrica**: Se valorará la originalidad en la solución aportada, así como la capacidad de discutir los resultados de forma detallada. El requisito mínimo servirá para aprobar la actividad, bajo premisa de que la discusión del resultado sera apropiada.

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).
* Se deberá entregar unicamente el notebook y los pesos del mejor modelo en un fichero .zip, de forma organizada.
* Cada alumno deberá de subir la solución de forma individual.

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [ ]:
# ==================== IMPORTS ====================

from __future__ import division

import os
import re
import glob
import json
import base64
import io
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from PIL import Image

import gym
from gym import wrappers
from pyvirtualdisplay import Display
from IPython.display import HTML

import keras
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from keras.optimizers import Adam
import keras.backend as K
from keras.callbacks import Callback

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger


#### Configuración

In [ ]:
# ==================== CONSTANTES GENERALES ====================
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4
ENV_NAME = 'SpaceInvaders-v0'
SEED = 123

In [ ]:
# ==================== PREPROCESAMIENTO Y UTILS ====================
class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        return batch.astype('float32') / 255.

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

def build_model(input_shape, nb_actions, conv_layers, dense_units):
    model = Sequential()
    if K.image_data_format() == 'channels_last':
        model.add(Permute((2, 3, 1), input_shape=input_shape))
    else:
        model.add(Permute((1, 2, 3), input_shape=input_shape))
    for filters, kernel_size, stride in conv_layers:
        model.add(Convolution2D(filters, (kernel_size, kernel_size), strides=(stride, stride), activation='relu'))
    model.add(Flatten())
    model.add(Dense(dense_units, activation='relu'))
    model.add(Dense(nb_actions, activation='linear'))
    return model


#### Callbacks y funciones auxiliares

In [ ]:
def plot_experiments(experiments, fields_to_plot):
    # Asignar un color distinto para cada experimento
    colors = plt.cm.tab10(np.linspace(0, 1, len(experiments)))
    data = {}
    for name, path in experiments.items():
        with open(path, 'r') as f:
            data[name] = json.load(f)
    dfs = {name: pd.DataFrame(exp_data) for name, exp_data in data.items()}
    n_fields = len(fields_to_plot)
    n_cols = min(n_fields, 3)
    n_rows = int(np.ceil(n_fields / n_cols))
    fig, axs = plt.subplots(n_rows, n_cols, figsize=(5 * n_cols, 4 * n_rows))
    axs = np.array(axs).flatten()
    for i, field in enumerate(fields_to_plot):
        for (name, df), color in zip(dfs.items(), colors):
            if field in df.columns:
                axs[i].plot(df[field], label=name, color=color)
        axs[i].set_xlabel("Timestep", fontsize=13)
        axs[i].set_ylabel(field, fontsize=13)
        axs[i].set_title(field, fontsize=14)
        axs[i].legend()
        axs[i].grid(True)
    for j in range(n_fields, len(axs)):
        fig.delaxes(axs[j])
    plt.tight_layout()
    plt.show()

def show_video():
    mp4list = glob.glob('./Videos/*.mp4')
    if not mp4list:
        print("No hay video.")
        return
    video_path = mp4list[0]
    with open(video_path, 'rb') as f:
        video = f.read()
    encoded = base64.b64encode(video).decode('ascii')
    return HTML(f'''
        <video width="480" height="auto" controls>
            <source src="data:video/mp4;base64,{encoded}" type="video/mp4" />
        </video>''')

In [ ]:
# ==================== CALLBACK CHECKPOINT ====================
class StatefulModelIntervalCheckpoint(Callback):

    def __init__(self, filepath, interval, initial_steps=0, verbose=1):
        super().__init__()
        self.filepath = filepath
        self.interval = interval
        self.verbose = verbose
        self.initial_steps = initial_steps
        self.last_saved_at_step = initial_steps
        self.current_fit_steps = 0

    def on_step_end(self, step, logs={}):
        self.current_fit_steps += 1
        absolute_step = self.initial_steps + self.current_fit_steps
        intervals_now = absolute_step // self.interval
        intervals_before = self.last_saved_at_step // self.interval
        if intervals_now > intervals_before:
            save_step = intervals_now * self.interval
            filepath = self.filepath.format(step=save_step)
            if self.verbose > 0:
                print(f'\n[Callback] Guardando pesos para el paso {save_step}.')
            self.model.save_weights(filepath, overwrite=True)
            self.last_saved_at_step = absolute_step

In [ ]:
# ==================== ENTRENAMIENTO Y TEST ====================
def find_latest_checkpoint(prefix):
    checkpoint_pattern = re.compile(f"^{prefix}_(\\d+)\\.h5f(\\.index)?$")
    latest_step = -1
    latest_weights_file = None
    for filename in os.listdir():
        match = checkpoint_pattern.match(filename)
        if match:
            step = int(match.group(1))
            if step > latest_step:
                latest_step = step
                latest_weights_file = filename.replace('.index', '')
    return latest_step, latest_weights_file


In [ ]:
def run_dqn_experiment(
    experiment_tag="baseline",
    memory_size=1000000,
    learning_rate=0.0001,
    annealing_steps=1000000,
    annealing_steps_warmup=50000,
    total_training_steps=1750000,
    gamma=0.99,
    target_model_update=10000,
    train_interval=4,
    delta_clip=1.0,
    enable_double_dqn=False,
    epsilon_max=1.0,
    epsilon_min=0.1,
    epsilon_test=0.05,
    conv_layers=[(32, 8, 4), (64, 4, 2), (64, 3, 1)],
    dense_units=512,
    test_episodes=10,
    batch_size=None,
    verbose_test=2,
    input_shape=(84, 84),
    window_length=4,
    env_name=ENV_NAME,
    seed=SEED
):
    print(f"\n========== EXPERIMENTO: {experiment_tag} ==========")
    print(f"Hiperparámetros: memory={memory_size}, lr={learning_rate}, annealing={annealing_steps}, double_dqn={enable_double_dqn}")

    env = gym.make(env_name)
    np.random.seed(seed)
    env.seed(seed)
    nb_actions = env.action_space.n

    net_input_shape = (window_length,) + input_shape
    model = build_model(net_input_shape, nb_actions, conv_layers, dense_units)
    print(model.summary())

    memory = SequentialMemory(limit=memory_size, window_length=window_length)
    policy = LinearAnnealedPolicy(
        EpsGreedyQPolicy(),
        attr='eps',
        value_max=epsilon_max,
        value_min=epsilon_min,
        value_test=epsilon_test,
        nb_steps=annealing_steps)
    processor = AtariProcessor()

    dqn = DQNAgent(
        model=model,
        nb_actions=nb_actions,
        policy=policy,
        memory=memory,
        processor=processor,
        nb_steps_warmup=annealing_steps_warmup,
        gamma=gamma,
        target_model_update=target_model_update,
        train_interval=train_interval,
        delta_clip=delta_clip,
        enable_double_dqn=enable_double_dqn
    )
    if batch_size is not None:
        dqn.batch_size = batch_size
    dqn.compile(Adam(learning_rate=learning_rate), metrics=['mae'])

    # Checkpoint y log filenames
    weights_filename = f'dqn_{experiment_tag}_{env_name}_weights.h5f'
    checkpoint_weights_filename = f'dqn_{experiment_tag}_{env_name}_weights_{{step}}.h5f'
    log_filename = f'dqn_{experiment_tag}_{env_name}_log.json'

    # Retoma de checkpoints
    prefix = f'dqn_{experiment_tag}_{env_name}_weights'
    completed_steps, latest_weights_file = find_latest_checkpoint(prefix)

    if latest_weights_file:
        print(f"Checkpoint encontrado: {latest_weights_file} (paso {completed_steps})")
        dqn.load_weights(latest_weights_file)
        fraction_completed = min(1.0, float(completed_steps) / float(annealing_steps))
        current_epsilon = epsilon_max - fraction_completed * (epsilon_max - epsilon_min)
        remaining_annealing_steps = annealing_steps - completed_steps
        print(f"Configurando política para reanudar: Epsilon inicial = {current_epsilon:.4f}, Pasos de decaimiento restantes = {remaining_annealing_steps}")
        final_policy = LinearAnnealedPolicy(
            EpsGreedyQPolicy(),
            attr='eps',
            value_max=current_epsilon,
            value_min=epsilon_min,
            value_test=epsilon_test,
            nb_steps=remaining_annealing_steps)
        dqn.policy = final_policy
        dqn.nb_steps_warmup = 32

    remaining_steps = total_training_steps - completed_steps
    callbacks = [StatefulModelIntervalCheckpoint(checkpoint_weights_filename, interval=25000, initial_steps=completed_steps, verbose=1)]
    callbacks += [FileLogger(log_filename, interval=100)]

    if remaining_steps > 0:
        print(f"\nEntrenando durante {remaining_steps} pasos para '{experiment_tag}'...")
        dqn.fit(env, callbacks=callbacks, nb_steps=remaining_steps, log_interval=10000)
        dqn.save_weights(weights_filename, overwrite=True)
    else:
        print("\nEl entrenamiento ya se había completado. No hay más pasos que entrenar.")

    # -- TEST --
    print("\nTesteando el modelo con los mejores pesos guardados...")
    dqn.load_weights(weights_filename)
    test_history = dqn.test(env, nb_episodes=test_episodes, visualize=False, verbose=verbose_test)
    rewards = test_history.history['episode_reward'] if 'episode_reward' in test_history.history else [ep['episode_reward'] for ep in test_history.history['episode']]
    mean_reward = np.mean(rewards)
    std_reward = np.std(rewards)
    print(f"\n[RESULTADOS {experiment_tag}]\nRecompensa media en test: {mean_reward:.2f}\nDesviación estándar: {std_reward:.2f}\nRecompensas individuales: {rewards}")

    return {
        'experiment': experiment_tag,
        'mean_reward': mean_reward,
        'std_reward': std_reward,
        'rewards': rewards,
        'params': dict(
            memory_size=memory_size,
            learning_rate=learning_rate,
            annealing_steps=annealing_steps,
            enable_double_dqn=enable_double_dqn,
            conv_layers=conv_layers,
            dense_units=dense_units
        )
    }


In [ ]:
def get_best_checkpoint_for_experiment(experiment_tag, env_name=ENV_NAME, checkpoint_dir="."):
    """
    Busca el checkpoint con mayor número de pasos para un experimento.
    Devuelve (completed_steps, checkpoint_file) o (None, None) si no encuentra.
    """
    prefix = f'dqn_{experiment_tag}_{env_name}_weights'
    checkpoint_pattern = re.compile(f"^{re.escape(prefix)}_(\\d+)\\.h5f(\\.index)?$")
    latest_step = -1
    latest_weights_file = None
    for filename in os.listdir(checkpoint_dir):
        match = checkpoint_pattern.match(filename)
        if match:
            step = int(match.group(1))
            if step > latest_step:
                latest_step = step
                latest_weights_file = os.path.join(checkpoint_dir, filename.replace('.index', ''))
    if latest_weights_file:
        return latest_step, latest_weights_file
    return None, None



def continue_training_with_improvements(
    base_experiment_tag,
    new_experiment_tag=None,
    additional_steps=250_000,
    new_learning_rate=None,
    new_epsilon_min=None,
    new_target_model_update=None,
    new_batch_size=None,
    checkpoint_dir=".",
    **kwargs
):
    """
    Continúa el entrenamiento de un experimento anterior desde su mejor checkpoint,
    aplicando mejoras/hyperparam tuning (por ejemplo: menor lr, menor epsilon_min, etc).
    """
    if new_experiment_tag is None:
        new_experiment_tag = f"{base_experiment_tag}_finetune"
    print(f"\n==== CONTINUANDO EXPERIMENTO: {base_experiment_tag} → {new_experiment_tag} ====")
    completed_steps, latest_weights_file = get_best_checkpoint_for_experiment(base_experiment_tag, ENV_NAME, checkpoint_dir)
    if latest_weights_file is None:
        print("No se encontró checkpoint previo, abortando.")
        return None
    print(f"Checkpoint encontrado: {latest_weights_file} (paso {completed_steps})")
    # Preparar hiperparámetros nuevos
    params = dict(
        experiment_tag=new_experiment_tag,
        env_name=ENV_NAME,
        seed=SEED,
    )
    params.update(kwargs)
    if new_learning_rate is not None:
        params['learning_rate'] = new_learning_rate
    if new_epsilon_min is not None:
        params['epsilon_min'] = new_epsilon_min
    if new_target_model_update is not None:
        params['target_model_update'] = new_target_model_update
    if new_batch_size is not None:
        params['batch_size'] = new_batch_size
    # Ajustar pasos de entrenamiento
    total_training_steps = completed_steps + additional_steps
    params['total_training_steps'] = total_training_steps
    # Llamar a run_dqn_experiment (asumiendo que toma en cuenta los checkpoints automáticamente)
    result = run_dqn_experiment(
        **params
    )
    return result


In [ ]:
# ==================== VIDEOS DE LAS PARTIDAS ====================

def record_video(agent, env_name, weights_file, video_folder="./Videos"):
    # Crea el entorno con monitor para grabar el vídeo
    env = gym.make(env_name)
    env = wrappers.Monitor(env, video_folder, force=True)
    agent.load_weights(weights_file)
    # Ejecuta una partida para grabar el vídeo
    agent.test(env, nb_episodes=1, visualize=False)
    env.close()


def show_video(video_folder="./Videos"):
    mp4list = glob.glob(f'{video_folder}/*.mp4')
    if len(mp4list) == 0:
        print("No hay video.")
        return
    video_path = mp4list[0]
    video = io.open(video_path, 'r+b').read()
    encoded = base64.b64encode(video)
    return HTML(data=f'''
        <video width="480" height="auto" controls>
            <source src="data:video/mp4;base64,{encoded.decode('ascii')}" type="video/mp4" />
        </video>''')



def show_best_checkpoint_video(experiment_tag, video_folder="./Videos", checkpoint_dir=".",
                               conv_layers=[(32, 8, 4), (64, 4, 2), (64, 3, 1)],
                               dense_units=512):
    """
    Graba y muestra el video usando el mejor checkpoint disponible para el experimento.
    """
    completed_steps, latest_weights_file = get_best_checkpoint_for_experiment(experiment_tag, ENV_NAME, checkpoint_dir)
    if not latest_weights_file:
        print("No hay checkpoint para este experimento.")
        return
    # Reconstruir agente con los mismos parámetros principales
    env = gym.make(ENV_NAME)
    nb_actions = env.action_space.n
    net_input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
    # Ajusta conv_layers y dense_units si lo necesitas, según tu experimento
    model = build_model(net_input_shape, nb_actions,
                        conv_layers=conv_layers,
                        dense_units=dense_units)
    memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
    policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps', value_max=1., value_min=0.1, value_test=0.05, nb_steps=1000000)
    processor = AtariProcessor()
    dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy, memory=memory, processor=processor, nb_steps_warmup=50000, gamma=0.99, target_model_update=10000, train_interval=4, delta_clip=1.0, enable_double_dqn=False)
    dqn.compile(Adam(learning_rate=0.0001), metrics=['mae'])
    dqn.load_weights(latest_weights_file)
    # Graba el video
    record_video(dqn, ENV_NAME, latest_weights_file, video_folder)
    # Muestra el video
    return show_video(video_folder)


---

#### **1. EXPERIMENTOS DQN VANILLA**

#### 1.1 Baseline DQN (punto de referencia)

Se realiza este experimento básico para tener un punto de partida para comparar con siguientes cambios

In [ ]:
results_vanilla_baseline = run_dqn_experiment(experiment_tag="vanilla_baseline", enable_double_dqn=False)



========== EXPERIMENTO: vanilla_baseline ==========
Hiperparámetros: memory=1000000, lr=0.0001, annealing=1000000, double_dqn=False
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_1 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_3 (Conv2D)           (None, 20, 20, 32)        8224      
                                                                 
 conv2d_4 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_5 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten_1 (Flatten)         (None, 3136)              0         
                                                                 
 dense_2 (Dense)             (None, 512)             

In [ ]:
video_vanilla_dqn_bl = show_best_checkpoint_video("vanilla_baseline")
display(video_vanilla_dqn_bl)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 17.000, steps: 829


#### 1.2 DQN con memoria reducida

Si el entorno no es muy complejo, una memoria más pequeña puede acelerar el entrenamiento y reducir el riesgo de que el agente aprenda sobre datos muy antiguos (menos varianza temporal).
Se espera un aprendizaje más rápido, pero quizá peor política si se queda sin experiencias suficientes.

In [ ]:
results_vanilla_mem200k = run_dqn_experiment(
    experiment_tag="vanilla_mem200k",
    memory_size=200000,
    enable_double_dqn=False
)



========== EXPERIMENTO: vanilla_mem200k ==========
Hiperparámetros: memory=200000, lr=0.0001, annealing=1000000, double_dqn=False
Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_5 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_12 (Conv2D)          (None, 20, 20, 32)        8224      
                                                                 
 conv2d_13 (Conv2D)          (None, 9, 9, 64)          32832     
                                                                 
 conv2d_14 (Conv2D)          (None, 7, 7, 64)          36928     
                                                                 
 flatten_5 (Flatten)         (None, 3136)              0         
                                                                 
 dense_10 (Dense)            (None, 512)               

In [ ]:
video_vanilla_mem200k = show_best_checkpoint_video("vanilla_mem200k")
display(video_vanilla_mem200k)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 10.000, steps: 761


#### 1.3 DQN con decaimiento rápido de epsilon (250k steps)


El agente explora menos tiempo, por lo que explota antes lo aprendido. Puede ayudar a la varianza si explora mucho y olvida lo bueno.
Se espera un aprendizaje más rápido, pero riesgo de atascarse en óptimos locales.

In [ ]:
results_vanilla_eps250k = run_dqn_experiment(
    experiment_tag="vanilla_eps250k",
    annealing_steps=250000,
    enable_double_dqn=False
)


In [ ]:
video_vanilla_eps = show_best_checkpoint_video("vanilla_eps250k")
display(video_vanilla_eps)

#### 1.4 DQN con red más pequeña (solo 2 capas conv, 256 en densa)

Menos parámetros puede resultar en menos sobreajuste y entrenamiento más rápido. A veces mejora la estabilidad, aunque el techo de rendimiento suele bajar.
Se espera un entrenamiento más estable y menor varianza, aunque posiblemente peor score máximo.

In [ ]:
results_vanilla_smallnet = run_dqn_experiment(
    experiment_tag="vanilla_smallnet",
    conv_layers=[(32, 8, 4), (64, 4, 2)],
    dense_units=256,
    enable_double_dqn=False
)



========== EXPERIMENTO: vanilla_smallnet ==========
Hiperparámetros: memory=1000000, lr=0.0001, annealing=1000000, double_dqn=False
Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_2 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_5 (Conv2D)           (None, 20, 20, 32)        8224      
                                                                 
 conv2d_6 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 flatten_2 (Flatten)         (None, 5184)              0         
                                                                 
 dense_4 (Dense)             (None, 256)               1327360   
                                                                 
 dense_5 (Dense)             (None, 6)               

In [ ]:
video_vanilla_smallnet = show_best_checkpoint_video("vanilla_smallnet",
                                                    conv_layers=[(32, 8, 4), (64, 4, 2)],
                                                    dense_units=256)
display(video_vanilla_smallnet)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 24.000, steps: 946


#### 1.5 DQN con tasa de aprendizaje alta (0.0005)

El agente aprende más agresivamente. Puede salir bien si la recompensa crece muy despacio.
Se espera una mayor velocidad de aprendizaje, pero riesgo de inestabilidad.

In [ ]:
results_vanilla_lr5 = run_dqn_experiment(
    experiment_tag="vanilla_lr5",
    learning_rate=0.0005,
    enable_double_dqn=False
)



========== EXPERIMENTO: vanilla_lr5 ==========
Hiperparámetros: memory=1000000, lr=0.0005, annealing=1000000, double_dqn=False
Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_2 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_6 (Conv2D)           (None, 20, 20, 32)        8224      
                                                                 
 conv2d_7 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_8 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten_2 (Flatten)         (None, 3136)              0         
                                                                 
 dense_4 (Dense)             (None, 512)               160

In [ ]:
video_vanilla_lr5 = show_best_checkpoint_video("vanilla_lr5")
display(video_vanilla_lr5)

#### **2. EXPERIMENTOS DOUBLE DQN**

#### 2.1 Double DQN Baseline

Se realiza para tener una base del double DQN y para ver si la reducción de sobreestimación mejora la media y la varianza respecto al baseline DQN

In [ ]:
results_double_baseline = run_dqn_experiment(
    experiment_tag="double_baseline",
    enable_double_dqn=True
)



========== EXPERIMENTO: double_baseline ==========
Hiperparámetros: memory=1000000, lr=0.0001, annealing=1000000, double_dqn=True
Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_7 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_18 (Conv2D)          (None, 20, 20, 32)        8224      
                                                                 
 conv2d_19 (Conv2D)          (None, 9, 9, 64)          32832     
                                                                 
 conv2d_20 (Conv2D)          (None, 7, 7, 64)          36928     
                                                                 
 flatten_7 (Flatten)         (None, 3136)              0         
                                                                 
 dense_14 (Dense)            (None, 512)               

In [ ]:
video_double_bl = show_best_checkpoint_video("double_baseline")
display(video_double_bl)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 8.000, steps: 455


#### 2.2 Double DQN con memoria reducida

Lo usaremos para combinar ambas mejoras: menor sobreestimación y memoria reciente.

In [ ]:
results = run_dqn_experiment(
    experiment_tag="double_mem200k",
    memory_size=200000,
    enable_double_dqn=True
)



========== EXPERIMENTO: double_mem200k ==========
Hiperparámetros: memory=200000, lr=0.0001, annealing=1000000, double_dqn=True
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 3136)              0         
                                                                 
 dense (Dense)               (None, 512)               1606

/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


10000/10000 [==============================] - 658s 66ms/step - reward: 0.0200
13 episodes - episode_reward: 15.077 [5.000, 24.000] - loss: 0.012 - mae: 3.405 - mean_q: 4.145 - mean_eps: 0.100 - ale.lives: 2.000

Interval 2 (10000 steps performed)
10000/10000 [==============================] - 645s 65ms/step - reward: 0.0197
14 episodes - episode_reward: 14.143 [8.000, 23.000] - loss: 0.014 - mae: 3.373 - mean_q: 4.102 - mean_eps: 0.100 - ale.lives: 2.099

Interval 3 (20000 steps performed)
 4997/10000 [=============>................] - ETA: 5:25 - reward: 0.0212
[Callback] Guardando pesos para el paso 1750000.
 5000/10000 [==============>...............] - ETA: 5:25 - reward: 0.0214done, took 1629.230 seconds

Testeando el modelo con los mejores pesos guardados...
Testing for 10 episodes ...
Episode 1: reward: 26.000, steps: 894
Episode 2: reward: 16.000, steps: 750
Episode 3: reward: 18.000, steps: 963
Episode 4: reward: 12.000, steps: 820
Episode 5: reward: 17.000, steps: 740
Episod

In [ ]:
video_double_mem200k = show_best_checkpoint_video("double_mem200k")
display(video_double_mem200k)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 19.000, steps: 1178


#### 2.3 Double DQN + decaimiento de ε rápido

Para combinar reducción de exploración con reducción de sobreestimación.


In [ ]:
results_double_eps250k = run_dqn_experiment(
    experiment_tag="double_eps250k",
    annealing_steps=250000,
    enable_double_dqn=True
)


In [ ]:
video_double_eps250k = show_best_checkpoint_video("double_eps250k")
display(video_double_eps250k)

#### 2.4 Double DQN con tasa de aprendizaje baja (0.00005)

En Double DQN suele ayudar a la estabilidad usar learning rates bajos, lo que puede ayudar a reducir la varianza.

In [ ]:
results_double_lr05 = run_dqn_experiment(
    experiment_tag="double_lr05",
    learning_rate=0.00005,
    enable_double_dqn=True
)


In [ ]:
# Reconstruir el modelo
nb_actions = 6  # SpaceInvaders-v0 tiene 6 acciones
net_input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = build_model(net_input_shape, nb_actions,
                    conv_layers=[(32, 8, 4), (64, 4, 2), (64, 3, 1)],
                    dense_units=512)

# Reconstruir componentes
memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
policy = EpsGreedyQPolicy(eps=0.05)
processor = AtariProcessor()

# Crear el agente DQN
dqn = DQNAgent(model=model,
               nb_actions=nb_actions,
               memory=memory,
               policy=policy,
               processor=processor,
               nb_steps_warmup=0,
               target_model_update=10000,
               gamma=0.99,
               train_interval=4,
               delta_clip=1.0,
               enable_double_dqn=True)

dqn.compile(Adam(learning_rate=0.00005), metrics=['mae'])

# Cargar pesos
checkpoint = "dqn_double_lr05_SpaceInvaders-v0_weights_1000000.h5f"
dqn.load_weights(checkpoint)

# Evaluar 10 episodios
env = gym.make(ENV_NAME)
history = dqn.test(env, nb_episodes=10, visualize=False)
rewards = history.history["episode_reward"]

# Mostrar resultados
print("Recompensas:", rewards)
print(f"Media: {np.mean(rewards):.2f} ± {np.std(rewards):.2f}")

Testing for 10 episodes ...
Episode 1: reward: 25.000, steps: 935
Episode 2: reward: 23.000, steps: 919
Episode 3: reward: 21.000, steps: 900
Episode 4: reward: 24.000, steps: 943
Episode 5: reward: 26.000, steps: 862
Episode 6: reward: 21.000, steps: 737
Episode 7: reward: 26.000, steps: 1062
Episode 8: reward: 23.000, steps: 967
Episode 9: reward: 16.000, steps: 619
Episode 10: reward: 15.000, steps: 646
Recompensas: [25.0, 23.0, 21.0, 24.0, 26.0, 21.0, 26.0, 23.0, 16.0, 15.0]
Media: 22.00 ± 3.66


In [ ]:
video_double_lr05 = show_best_checkpoint_video("double_lr05")
display(video_double_lr05)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 25.000, steps: 1160


#### 2.5 Double DQN con batch size grande (64)

Este experimento lo realizamos porque un batch grande suaviza el gradiente y puede ayudar a reducir la varianza del aprendizaje.

In [ ]:
results_double_bs64 = run_dqn_experiment(
    experiment_tag="double_bs64",
    batch_size=64,
    enable_double_dqn=True
)


In [ ]:
# Reconstruir el modelo
nb_actions = 6  # SpaceInvaders-v0 tiene 6 acciones
net_input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = build_model(net_input_shape, nb_actions,
                    conv_layers=[(32, 8, 4), (64, 4, 2), (64, 3, 1)],
                    dense_units=512)

# Reconstruir componentes
memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
policy = EpsGreedyQPolicy(eps=0.05)
processor = AtariProcessor()

# Crear el agente DQN
dqn = DQNAgent(model=model,
               nb_actions=nb_actions,
               memory=memory,
               policy=policy,
               processor=processor,
               nb_steps_warmup=0,
               target_model_update=10000,
               gamma=0.99,
               train_interval=4,
               delta_clip=1.0,
               enable_double_dqn=True)

dqn.compile(Adam(learning_rate=0.00005), metrics=['mae'])

# Cargar pesos del experimento 'double_bs64'
checkpoint = "dqn_double_bs64_SpaceInvaders-v0_weights_1000000.h5f"
dqn.load_weights(checkpoint)

# Evaluar 10 episodios
env = gym.make(ENV_NAME)
history = dqn.test(env, nb_episodes=10, visualize=False)
rewards = history.history["episode_reward"]

# Mostrar resultados
print("Recompensas:", rewards)
print(f"Media: {np.mean(rewards):.2f} ± {np.std(rewards):.2f}")


Testing for 10 episodes ...
Episode 1: reward: 7.000, steps: 650
Episode 2: reward: 7.000, steps: 699
Episode 3: reward: 12.000, steps: 1133
Episode 4: reward: 7.000, steps: 597
Episode 5: reward: 7.000, steps: 605
Episode 6: reward: 16.000, steps: 1042
Episode 7: reward: 6.000, steps: 481
Episode 8: reward: 12.000, steps: 1164
Episode 9: reward: 4.000, steps: 597
Episode 10: reward: 16.000, steps: 1246
Recompensas: [7.0, 7.0, 12.0, 7.0, 7.0, 16.0, 6.0, 12.0, 4.0, 16.0]
Media: 9.40 ± 4.05


In [ ]:
video_double_bs64 = show_best_checkpoint_video("double_bs64")
display(video_double_bs64)

Testing for 1 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 10.000, steps: 879


#### **3. FINE TUNING DESDE LOS MEJORES CHECKPOINTS**

#### 3.1 Vanilla dqn baseline

In [ ]:
result_finetune_vanilla_baseline = continue_training_with_improvements(
    base_experiment_tag="vanilla_baseline",
    new_experiment_tag="vanilla_baseline_finetune",
    additional_steps=250_000,
    new_learning_rate=0.00002,
    new_epsilon_min=0.02,
    new_target_model_update=5000,
    new_batch_size=64
)


==== CONTINUANDO EXPERIMENTO: vanilla_baseline → vanilla_baseline_finetune ====
Checkpoint encontrado: ./dqn_vanilla_baseline_SpaceInvaders-v0_weights_1750000.h5f (paso 1750000)

========== EXPERIMENTO: vanilla_baseline_finetune ==========
Hiperparámetros: memory=1000000, lr=2e-05, annealing=1000000, double_dqn=False
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_1 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_3 (Conv2D)           (None, 20, 20, 32)        8224      
                                                                 
 conv2d_4 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_5 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                

In [ ]:
video_vanilla_bl_ft = show_best_checkpoint_video("vanilla_baseline_finetune")
display(video_vanilla_bl_ft)

#### 3.2 Double dqn baseline

In [ ]:
result_finetune_double_baseline = continue_training_with_improvements(
    base_experiment_tag="double_baseline",
    new_experiment_tag="double_baseline_finetune",
    additional_steps=250_000,
    new_learning_rate=0.00002,    # Aprendizaje menor
    new_epsilon_min=0.02,         # Menos exploración en fine-tuning
    new_target_model_update=5000, # Target más frecuente
    new_batch_size=64             # Batch más grande
)

In [ ]:
video_double_ft = show_best_checkpoint_video("double_baseline_finetune")
display(video_double_bl)

#### 3.3 Double dqn con memoria reducida

In [ ]:
result_finetune_double_mem200k = continue_training_with_improvements(
    base_experiment_tag="double_mem200k",
    new_experiment_tag="double_mem200k_finetune",
    additional_steps=250_000,
    new_learning_rate=0.00002,
    new_epsilon_min=0.02,
    new_target_model_update=5000,
    new_batch_size=64
)


==== CONTINUANDO EXPERIMENTO: double_mem200k → double_mem200k_finetune ====
Checkpoint encontrado: ./dqn_double_mem200k_SpaceInvaders-v0_weights_1750000.h5f (paso 1750000)

========== EXPERIMENTO: double_mem200k_finetune ==========
Hiperparámetros: memory=1000000, lr=2e-05, annealing=1000000, double_dqn=False
Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_2 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_6 (Conv2D)           (None, 20, 20, 32)        8224      
                                                                 
 conv2d_7 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_8 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatt

In [ ]:
video_double_mem_ft = show_best_checkpoint_video("double_mem200k_finetune")
display(video_double_mem_ft)

#### 3.4 Double DQN con tasa de aprendizaje baja (0.00005)

In [ ]:
from google.colab import files
uploaded = files.upload()
with open("evaluate_checkpoint_resultados_double_lr05.json", "r") as f:
    resultados = json.load(f)

# Mostrar resultados ordenadamente
for r in resultados:
    print(f"{r['file']} → Media: {r['mean_reward']:.2f} ± {r['std_reward']:.2f}")

Saving result_finetune_manual(1).json to result_finetune_manual(1).json
Saving result_finetune_manual.json to result_finetune_manual (1).json
dqn_double_lr05_SpaceInvaders-v0_weights_1000000.h5f → Media: 20.90 ± 5.58
dqn_double_lr05_SpaceInvaders-v0_weights_50000.h5f → Media: 20.70 ± 9.69
dqn_double_lr05_SpaceInvaders-v0_weights_925000.h5f → Media: 19.60 ± 5.00
dqn_double_lr05_SpaceInvaders-v0_weights_725000.h5f → Media: 18.70 ± 3.07
dqn_double_lr05_SpaceInvaders-v0_weights_775000.h5f (1) → Media: 18.50 ± 3.77
dqn_double_lr05_SpaceInvaders-v0_weights_25000.h5f (1) → Media: 18.20 ± 6.63
dqn_double_lr05_SpaceInvaders-v0_weights_650000.h5f → Media: 18.10 ± 7.02
dqn_double_lr05_SpaceInvaders-v0_weights.h5f (2) → Media: 18.00 ± 6.13
dqn_double_lr05_SpaceInvaders-v0_weights.h5f → Media: 18.00 ± 2.61
dqn_double_lr05_SpaceInvaders-v0_weights_675000.h5f → Media: 17.40 ± 5.02
dqn_double_lr05_SpaceInvaders-v0_weights_850000.h5f (1) → Media: 17.40 ± 6.71
dqn_double_lr05_SpaceInvaders-v0_weights_92

In [ ]:
result_finetune_double_lr05 = continue_training_with_improvements(
    base_experiment_tag="double_lr05",
    new_experiment_tag="double_lr05_finetune",
    additional_steps=250_000,
    new_learning_rate=0.00002,
    new_epsilon_min=0.02,
    new_target_model_update=5000,
    new_batch_size=64
)


==== CONTINUANDO EXPERIMENTO: double_lr05 → double_lr05_finetune ====
Checkpoint encontrado: ./dqn_double_lr05_SpaceInvaders-v0_weights_1000000.h5f (paso 1000000)

========== EXPERIMENTO: double_lr05_finetune ==========
Hiperparámetros: memory=1000000, lr=2e-05, annealing=1000000, double_dqn=False
Model: "sequential_48"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_48 (Permute)        (None, 84, 84, 4)         0         
                                                                 
 conv2d_144 (Conv2D)         (None, 20, 20, 32)        8224      
                                                                 
 conv2d_145 (Conv2D)         (None, 9, 9, 64)          32832     
                                                                 
 conv2d_146 (Conv2D)         (None, 7, 7, 64)          36928     
                                                                 
 flatten_48 (Flat



---



#### **4. EVALUACIÓN EN TEST DE LOS MEJORES 3 MODELOS.**

In [ ]:
# Reconstruir el modelo
nb_actions = env.action_space.n
net_input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = build_model(net_input_shape, nb_actions,
                    conv_layers=[(32, 8, 4), (64, 4, 2), (64, 3, 1)],
                    dense_units=512)

# Reconstruir componentes
memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
policy = EpsGreedyQPolicy()
processor = AtariProcessor()

# Crear el agente DQN
dqn = DQNAgent(model=model,
               nb_actions=nb_actions,
               memory=memory,
               policy=policy,
               processor=processor,
               nb_steps_warmup=0,
               target_model_update=10000,
               gamma=0.99,
               train_interval=4,
               delta_clip=1.0,
               enable_double_dqn=True)

dqn.compile(Adam(learning_rate=0.00005), metrics=['mae'])

env = gym.make(ENV_NAME)

# Double DQN con tasa de aprendizaje 0.00005
print("*****Testeando Double DQN con tasa de aprendizaje 0.00005*****")
dqn.load_weights('dqn_double_lr05_SpaceInvaders-v0_weights.h5f')
test_history = dqn.test(env, nb_episodes=10, visualize=False, verbose=True)
rewards = test_history.history['episode_reward'] if 'episode_reward' in test_history.history else [ep['episode_reward'] for ep in test_history.history['episode']]

print(f"\n[RESULTADOS Double DQN con tasa de aprendizaje 0.00005]\nRecompensa media en test: {np.mean(rewards):.2f}\nDesviación estándar: {np.std(rewards):.2f}\nRecompensas individuales: {rewards}")
print("\n-------------------------------\n")

# Vanilla DQN baseline
print("*****Testeando Vanilla DQN baseline*****")
dqn.load_weights('dqn_vanilla_baseline_SpaceInvaders-v0_weights.h5f')
test_history = dqn.test(env, nb_episodes=10, visualize=False, verbose=True)
rewards = test_history.history['episode_reward'] if 'episode_reward' in test_history.history else [ep['episode_reward'] for ep in test_history.history['episode']]

print(f"\n[RESULTADOS Vanilla DQN baseline]\nRecompensa media en test: {np.mean(rewards):.2f}\nDesviación estándar: {np.std(rewards):.2f}\nRecompensas individuales: {rewards}")
print("\n-------------------------------\n")

# Double DQN con tasa de aprendizaje 0.00005 fine-tuned
print("*****Testeando Double DQN con tasa de aprendizaje 0.00005 fine-tuned*****")
dqn.load_weights('dqn_double_lr05_SpaceInvaders-v0_weights_1000000.h5f')
test_history = dqn.test(env, nb_episodes=10, visualize=False, verbose=True)
rewards = test_history.history['episode_reward'] if 'episode_reward' in test_history.history else [ep['episode_reward'] for ep in test_history.history['episode']]

print(f"\n[RESULTADOS Double DQN con tasa de aprendizaje 0.00005 fine-tuned]\nRecompensa media en test: {np.mean(rewards):.2f}\nDesviación estándar: {np.std(rewards):.2f}\nRecompensas individuales: {rewards}")
print("\n-------------------------------\n")


*****Testeando Double DQN con tasa de aprendizaje 0.0005*****
Testing for 10 episodes ...


/usr/local/lib/python3.11/dist-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


Episode 1: reward: 15.000, steps: 699
Episode 2: reward: 28.000, steps: 1144
Episode 3: reward: 11.000, steps: 532
Episode 4: reward: 20.000, steps: 900
Episode 5: reward: 31.000, steps: 1216
Episode 6: reward: 24.000, steps: 1058
Episode 7: reward: 22.000, steps: 949
Episode 8: reward: 13.000, steps: 713
Episode 9: reward: 18.000, steps: 828
Episode 10: reward: 14.000, steps: 661

[RESULTADOS Double DQN con tasa de aprendizaje 0.0005]
Recompensa media en test: 19.60
Desviación estándar: 6.31
Recompensas individuales: [15.0, 28.0, 11.0, 20.0, 31.0, 24.0, 22.0, 13.0, 18.0, 14.0]

-------------------------------

*****Testeando Vanilla DQN baseline*****
Testing for 10 episodes ...
Episode 1: reward: 18.000, steps: 903
Episode 2: reward: 22.000, steps: 892
Episode 3: reward: 16.000, steps: 685
Episode 4: reward: 29.000, steps: 846
Episode 5: reward: 17.000, steps: 803
Episode 6: reward: 21.000, steps: 850
Episode 7: reward: 19.000, steps: 778
Episode 8: reward: 28.000, steps: 1196
Episode

---
## **PARTE 4.** Justificación de los parámetros seleccionados y de los resultados obtenidos

#### 1. Introducción y enfoque del proyecto

Este proyecto se centra en resolver el entorno SpaceInvaders-v0 utilizando Deep Q-Learning (DQN), una técnica model-free del Aprendizaje por Refuerzo. Inspirados por el enfoque sistemático presentado en clase, se aplicaron principios de planificación y mejora incremental como el uso de checkpoints, pruebas controladas y fine-tuning, acercando el proyecto a una perspectiva model-based en la práctica experimental.

#### 2. Funciones clave desarrolladas
Con el fin de mantener un diseño modular, transparente y reutilizable, se implementaron diversas funciones clave:

•	build_model().
 Construcción del modelo CNN con capas convolucionales y densas, basado en el enfoque de DeepMind para Atari.

•	AtariProcessor.
Procesamiento de imágenes (escalado a 84x84, escala de grises, normalización), siguiendo el preprocesamiento visto en clase.

•	run_dqn_experiment().
Función principal de entrenamiento y evaluación que está formada por:

    o	Reanudación automática desde el mejor checkpoint encontrado.

    o	Ajuste dinámico de hiperparámetros.

    o	Guardado periódico de pesos con StatefulModelIntervalCheckpoint.

•	continue_training_with_improvements().
Extensión de entrenamiento (fine-tuning) desde un experimento base, con nuevos
hiperparámetros.

•	get_best_checkpoint_for_experiment().
Localiza automáticamente el mejor checkpoint por pasos completados.

•	show_best_checkpoint_video() y plot_experiments().
Visualización de resultados y grabación del rendimiento del agente.

Estas funciones nos han permitido llevar a cabo múltiples ciclos experimentales de forma estructurada, lo cual ha sido clave en el análisis comparativo y la mejora progresiva.

#### 3. Justificación de los parámetros seleccionados
A continuación, se describen los principales hiperparámetros utilizados, junto con su justificación basada en teoría y experimentación:


•	INPUT_SHAPE (84, 84):
Reducción de resolución estándar en Atari que implica menor coste computacional.

•	WINDOW_LENGTH 4:
Captura de la dinámica temporal (4 frames consecutivos).

•	memory_size 1,000,000:
Almacenamiento suficiente sin saturar la RAM.

•	epsilon_min 0.02 – 0.1:
Rango reducido en fine-tune para mayor explotación.

•	learning_rate 0.00005 – 0.00002:
Reducción progresiva para evitar oscilaciones y afinar la política aprendida.

•	target_model_update 5000 – 10000:
Actualizaciones más frecuentes en fine-tune para una adaptación más rápida.

•	enable_double_dqn True:
Mejora el rendimiento al mitigar el overestimation bias.

•	batch_size 32 – 64:
Fine-tuning con mayor batch, aportó más estabilidad.

•	annealing_steps 1,000,000:
Desaceleración gradual de la exploración (epsilon).

Estos valores no solo se escogieron por su validez teórica, sino también tras comparar múltiples configuraciones a través de experimentación, lo cual se tradujo en mejoras concretas en el rendimiento del agente.

#### 4. Resultados obtenidos y análisis comparativo
A lo largo del proyecto se realizaron varios experimentos controlados, con el objetivo de evaluar el impacto de diferentes configuraciones sobre el rendimiento del agente. Los resultados obtenidos permiten identificar patrones claros de mejora y reflexionar sobre las decisiones tomadas.

El mejor rendimiento global se alcanzó con el experimento double_lr05_finetune, donde se retomó el entrenamiento desde el checkpoint de 1.000.000 pasos. Este fine-tune bien dirigido elevó la recompensa media a 22.3 puntos, confirmando que continuar el entrenamiento a partir de un modelo ya entrenado puede mejorar sustancialmente el desempeño, especialmente si se hace con hiperparámetros refinados.

También se evaluaron agentes sin fine-tuning. El Double DQN base, entrenado con una tasa de aprendizaje de 0.00005, logró una media de 19.6, mostrando ya un progreso considerable respecto a versiones sin Double DQN. Una observación especialmente relevante surgió al analizar individualmente los checkpoints, ya que en el archivo correspondiente a solo 50.000 pasos de entrenamiento alcanzó una recompensa media de 20.7, superando incluso a varios modelos más entrenados. Este hallazgo sugiere que, en determinadas configuraciones, el agente puede aprender patrones útiles de forma temprana, aunque no siempre logre mantenerlos o mejorarlos con más pasos.

 Curiosamente, el Vanilla DQN baseline, sin ninguna técnica avanzada, obtuvo 21.0 puntos, lo que lo posiciona como una línea base sólida y útil para comparación.

Estos resultados, evaluados tanto globalmente como por checkpoints, refuerzan la importancia de una exploración iterativa y modular. La estructura del código nos facilitó una reanudación flexible desde cualquier punto del entrenamiento, permitiendo validar con rigor distintas hipótesis a lo largo del proceso.

#### 5. Extensiones pendientes y potencial de mejora
Durante el desarrollo del proyecto se priorizaron configuraciones con mayor viabilidad de mejora y validación empírica. Si bien no fue posible completar todas las combinaciones experimentales por cuestiones de tiempo y recursos, el proyecto deja preparada la base para una fácil extensión. Por ejemplo:

•	El experimento double_epsilon001, diseñado para explorar una política más explotativa, está definido y listo para ejecutar.

•	Combinaciones interesantes como menor memoria de replay junto con batch reducido quedaron identificadas para futuras pruebas.
Como ya se ha comentado, gracias al diseño modular y reutilizable del código, todos estos escenarios pueden ser retomados de forma inmediata, permitiendo continuar el ciclo de mejora progresiva sin rehacer el flujo de trabajo.
